src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+    if prag.name == "table" and prag.kind == pkKval:
+      return $prag.value
+
+
+proc getCollectionName*(T: typedesc): string =
+  ##[ Get the name of the DB table for the given type: ``table`` pragma value if it exists
+  or lowercased type name otherwise.
+  ]##
+
+  when T.hasCustomPragma(table): T.getCustomPragmaVal(table)
+  else: ($T).toLowerAscii()  
+
+
+var
+  allCollections*: seq[string] = @[]
+  normConnection*: string
+  normUser*: string
+  normPassword*: string
+  normDatabase*: string
+
+template genWithoutDb(newCollections: seq[string]): untyped {.dirty.} =
+  ## simply add to global allCollections
+
+  allCollections.insert(newCollections)
+
+template genWithDb(connection, user, password, database: string, useParams: int, newCollections: seq[string]): untyped {.dirty.} =
+  ## Generate ``withDb`` templates.
+
+  allCollections.insert(newCollections)
+  if useParams == -1:
+    normConnection = connection
+    normUser = user
+    normPassword = password
+    normDatabase = database
+
+  template withCustomDb*(
+    customConnection, customUser, customPassword, customDatabase: string,
+    body: untyped
+  ): untyped {.dirty.} =
+    ##[ A wrapper for actions that require custom DB connection, i.e. not the one defined in ``db``.
+    Defines CRUD procs to work with the DB.
+
+    'connection' should contain the URI pointing to the MongoDB server.
+
+    The 'user' and 'password' parameters are not used.
+
+    'database' should contain the database shard.
+
+    Aforementioned procs and procs defined in a ``db`` block can be used only
+    in  a ``withDb`` block.
+    ]##
+
+    block:
+      var dbConn = newMongoWithURI(customConnection)
+      let
+        dbConnResult = dbConn.connect()  ## TODO: Add more error handling
+
+      template dropTables() {.used.} =
+        ## Drops the collections for ALL objects.
+
+        for c in allCollections:
+          let dbCollection = dbConn[customDatabase][c]
+          let removeResult = dbCollection.drop()
+
+      template createTables(force = false) {.used.} =
+        ##[ For other database types, this function create new tables. However,
+        for MongoDB this has less meaning since collections are auto-created.
+        However, if `force` is set to true, it will dropTables first, deleting all
+        the documents.
+        ]##
+
+        if force:
+          dropTables()
+
+      proc insert(obj: var object, force = false) {.used.} =
+        ##[ Insert object instance as a document into DB.The object's id is updated after
+        the insertion.
+
+        The ``force`` parameter is ignored by the mongodb library.
+        ]##
+
+        let
+          doc = toBson(obj, force)
+          dbCollection = dbConn[customDatabase][getCollectionName(type(obj))]
+
+        var response = dbCollection.insert(doc)
+
+        if len(response.inserted_ids) > 0:
+          obj.id = response.inserted_ids[0].toOid
+
+      proc pullOne(obj: var object, id: Oid) {.used.} =

```

Comment:
What is the purpose o renaming getOne to pullOne? The original idea is that you have a getOne that you can use with both instances and types, which, I thought, was convenient,

Reply:
I can see how that makes things more explicit, but I'd like to hear your reasoning.

Reply:
As mentioned in a earlier reply, I found the same function having different effects on the first parameter confusing. But I think it is better to be the same as the other databases, so I am reverting it all back to "get".